### ISSUE TITLE
Raiden stops when settling a channel fails

### ISSUE THREAD

**Issue by rakanalh** [raiden-network/raiden#1943]
## Problem Definition

When starting a node which uses an address that participated in a channel, the node starts to sync with the blockchain and then it stops because the settle transaction fails:
```
2018-07-25 09:13:40 [error    ] UnlockFailed!                  [raiden.raiden_event_handler] reason=lock expired secrethash=66d206ff
2018-07-25 09:13:40 [error    ] UnlockFailed!                  [raiden.raiden_event_handler] reason=lock expired secrethash=dce9fabf
2018-07-25 09:13:40 [info     ] close called                   [raiden.network.proxies.token_network] additional_hash=0x balance_hash=0x node=fc57d325 nonce=0 partner=774afb06 signature=0x token_network=7692624c
2018-07-25 09:13:41 [info     ] Channel with 774afb06 for token 0f114a1e is already closed [raiden.raiden_event_handler]
2018-07-25 09:13:41 [info     ] close called                   [raiden.network.proxies.token_network] additional_hash=0xdd8db5686a977633b14ee94e6782198b8043dd19ef490a8e432189cd2836b932 balance_hash=0x48fadce810413ca37d082c81328d0a3af41787c02f6d8bd8c7967451b0c0d4f2 node=fc57d325 nonce=5 partner=92746ece signature=0x5444c55b2589cf4c4e7f9c7b7d94cde16710d57947bcd69149caeda86b18a80e6b7630617552c2490a776534ed8e169f6037b7ba244a1595c2f257510dae53211b token_network=7692624c
2018-07-25 09:13:41 [info     ] Channel with 92746ece for token 0f114a1e is already closed [raiden.raiden_event_handler]
2018-07-25 09:13:41 [info     ] close called                   [raiden.network.proxies.token_network] additional_hash=0x00735598a02f4071427eca0817c93021e788a91d1822b5721d4ae95dab0c863f balance_hash=0x2016ab3033095829ecb0ade07bddb04edf6ba73b18922a1f1dc915e51d77e64f node=fc57d325 nonce=4 partner=9d44849d signature=0xde2e8aa3f2eace0089912c4aa5d734da3ca388c32091d886500b91d1a590f5524a1ca238f41e3bf742bc613b9fc4d300b0d2f0f143ebbca4d692ba28891baa131b token_network=7692624c
2018-07-25 09:13:41 [info     ] Channel with 9d44849d for token 0f114a1e is already closed [raiden.raiden_event_handler]
2018-07-25 09:13:41 [info     ] close called                   [raiden.network.proxies.token_network] additional_hash=0x balance_hash=0x node=fc57d325 nonce=0 partner=82641569 signature=0x token_network=7692624c
2018-07-25 09:13:41 [info     ] Channel with 82641569 for token 0f114a1e is already closed [raiden.raiden_event_handler]
2018-07-25 09:13:41 [info     ] settle called                  [raiden.network.proxies.token_network] locked_amount=12 locksroot=0x4c77302de4fbc3517b53d7bb1549ec91df9d9c6c250952a36fd5dbc55417b312 node=fc57d325 partner=92746ece partner_locked_amount=11 partner_locksroot=0x784569b0830aea7c3292847f1860f5993d96538851ca89980a2b95069116886e partner_transferred_amount=12 token_network=7692624c transferred_amount=0
2018-07-25 09:13:46 [debug    ] channel data {'channel_identifier': b'\x83\xfd&\x13\xe5\xfc\x9d8{#\x8c\x8b\xbe\xfd\xcb\x1b\t\x02\xf6&\xb0\xbb\xf0]\x08\xf1\xa9d[\x98\x93\x89', 'settle_block_number': 3700856, 'state': 2} [raiden.network.proxies.token_network]
2018-07-25 09:13:46 [info     ] settle failed                  [raiden.network.proxies.token_network] locked_amount=12 locksroot=0x4c77302de4fbc3517b53d7bb1549ec91df9d9c6c250952a36fd5dbc55417b312 node=fc57d325 partner=92746ece partner_locked_amount=11 partner_locksroot=0x784569b0830aea7c3292847f1860f5993d96538851ca89980a2b95069116886e partner_transferred_amount=12 token_network=7692624c transferred_amount=0
FATAL: Settle transaction threw. Receipt=AttributeDict({'blockHash': HexBytes('0x9ca24c1382ce18d17f890506add65effbec34848ef7cbf842fe753a9abfbd74d'), 'blockNumber': 3707389, 'contractAddress': None, 'cumulativeGasUsed': 34474, 'from': '0xfc57d325f23b9121a8488ffde2e6b3ef1208a20b', 'gasUsed': 34474, 'logs': [], 'logsBloom': HexBytes('0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'), 'status': 0, 'to': '0x7692624c0c43285d51c0dddfb2a6388f5002762e', 'transactionHash': HexBytes('0xc7ab7f7dbb24ffde5bca90f0704e338362ed7e95381ec0c9ca4c18523da9b11c'), 'transactionIndex': 0})
```

As you can notice from the above logs, i have triggered multiple channels to close with some succeeding in settling and the last one failed. The logs also contain secret expired error which **probably** caused settling to fail.

**Comment by pcppcp** [comment]
For me this happens with every run now (even with a clean db).

https://gist.github.com/bb053b5b31f5810cda1bf949eaa6adb3

**Comment by LefterisJP** [comment]
Can you check the blockchain by calling the TokenNetwork contract with both participant addresses either from python or from geth and querying [getChannelIdentifier()](https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/contracts/TokenNetwork.sol#L842) function?

To do that from geth simply do the following. I assume you are using the RTT token whose `TokenNetwork` contract is [this](https://ropsten.etherscan.io/address/0x7692624c0c43285d51c0dddfb2a6388f5002762e#code). If not, just adjust the token network address.

Add this into a .js file

```javascript
var abi = [{"constant":false,"inputs":[{"name":"participant1","type":"address"},{"name":"participant2","type":"address"},{"name":"settle_timeout","type":"uint256"}],"name":"openChannel","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"settlement_timeout_max","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"secret_registry","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"participant","type":"address"},{"name":"partner","type":"address"},{"name":"merkle_tree_leaves","type":"bytes"}],"name":"unlock","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"chain_id","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"participant","type":"address"},{"name":"partner","type":"address"},{"name":"locksroot","type":"bytes32"}],"name":"getLocksrootIdentifier","outputs":[{"name":"key","type":"bytes32"}],"payable":false,"stateMutability":"pure","type":"function"},{"constant":true,"inputs":[],"name":"channel_counter","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"MAX_SAFE_UINT256","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"contract_address","type":"address"}],"name":"contractExists","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"participant1","type":"address"},{"name":"participant1_transferred_amount","type":"uint256"},{"name":"participant1_locked_amount","type":"uint256"},{"name":"participant1_locksroot","type":"bytes32"},{"name":"participant2","type":"address"},{"name":"participant2_transferred_amount","type":"uint256"},{"name":"participant2_locked_amount","type":"uint256"},{"name":"participant2_locksroot","type":"bytes32"}],"name":"settleChannel","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"}],"name":"channels","outputs":[{"name":"settle_block_number","type":"uint256"},{"name":"state","type":"uint8"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"participant","type":"address"},{"name":"partner","type":"address"}],"name":"getParticipantsHash","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"pure","type":"function"},{"constant":false,"inputs":[{"name":"participant1_address","type":"address"},{"name":"participant1_balance","type":"uint256"},{"name":"participant2_address","type":"address"},{"name":"participant2_balance","type":"uint256"},{"name":"participant1_signature","type":"bytes"},{"name":"participant2_signature","type":"bytes"}],"name":"cooperativeSettle","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"participant","type":"address"},{"name":"partner","type":"address"}],"name":"getChannelIdentifier","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"partner","type":"address"},{"name":"balance_hash","type":"bytes32"},{"name":"nonce","type":"uint256"},{"name":"additional_hash","type":"bytes32"},{"name":"signature","type":"bytes"}],"name":"closeChannel","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"name":"participant","type":"address"},{"name":"total_deposit","type":"uint256"},{"name":"partner","type":"address"}],"name":"setTotalDeposit","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"participant","type":"address"},{"name":"partner","type":"address"}],"name":"getChannelParticipantInfo","outputs":[{"name":"","type":"uint256"},{"name":"","type":"uint256"},{"name":"","type":"bool"},{"name":"","type":"bytes32"},{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"closing_participant","type":"address"},{"name":"non_closing_participant","type":"address"},{"name":"balance_hash","type":"bytes32"},{"name":"nonce","type":"uint256"},{"name":"additional_hash","type":"bytes32"},{"name":"closing_signature","type":"bytes"},{"name":"non_closing_signature","type":"bytes"}],"name":"updateNonClosingBalanceProof","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"contract_version","outputs":[{"name":"","type":"string"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"settlement_timeout_min","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"participant","type":"address"},{"name":"total_withdraw","type":"uint256"},{"name":"partner","type":"address"},{"name":"participant_signature","type":"bytes"},{"name":"partner_signature","type":"bytes"}],"name":"setTotalWithdraw","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"}],"name":"participants_hash_to_channel_counter","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"deposit_limit","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"participant1","type":"address"},{"name":"participant2","type":"address"}],"name":"getChannelInfo","outputs":[{"name":"","type":"bytes32"},{"name":"","type":"uint256"},{"name":"","type":"uint8"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"token","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"participant1","type":"address"},{"name":"participant2","type":"address"},{"name":"locksroot","type":"bytes32"}],"name":"getParticipantLockedAmount","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"inputs":[{"name":"_token_address","type":"address"},{"name":"_secret_registry","type":"address"},{"name":"_chain_id","type":"uint256"},{"name":"_settlement_timeout_min","type":"uint256"},{"name":"_settlement_timeout_max","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":true,"name":"participant1","type":"address"},{"indexed":true,"name":"participant2","type":"address"},{"indexed":false,"name":"settle_timeout","type":"uint256"}],"name":"ChannelOpened","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":true,"name":"participant","type":"address"},{"indexed":false,"name":"total_deposit","type":"uint256"}],"name":"ChannelNewDeposit","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":true,"name":"participant","type":"address"},{"indexed":false,"name":"total_withdraw","type":"uint256"}],"name":"ChannelWithdraw","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":true,"name":"closing_participant","type":"address"}],"name":"ChannelClosed","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"participant","type":"address"},{"indexed":true,"name":"partner","type":"address"},{"indexed":true,"name":"locksroot","type":"bytes32"},{"indexed":false,"name":"unlocked_amount","type":"uint256"},{"indexed":false,"name":"returned_tokens","type":"uint256"}],"name":"ChannelUnlocked","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":true,"name":"closing_participant","type":"address"}],"name":"NonClosingBalanceProofUpdated","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"channel_identifier","type":"bytes32"},{"indexed":false,"name":"participant1_amount","type":"uint256"},{"indexed":false,"name":"participant2_amount","type":"uint256"}],"name":"ChannelSettled","type":"event"}];
var address = "0x7692624c0C43285D51c0DdDfb2A6388f5002762E";
var token_network = web3.eth.contract(abi).at(address);

var id = token_network.getChannelIdentifier("0xPARTICIPANT_ADDRESS", "0xPARTNERADDRESS");
console.log("Channel ID is: " + id);
```

From the geth console do: "loadScript("/path/to/thejsfileoucreated.js");

Compared the output channel ID with the one in the logs. If it's not the same then that means the channel was already recreated and it's the same issue as #1907 and the solution should be [this](https://github.com/raiden-network/raiden/issues/1907#issuecomment-407565857).

**Comment by pcppcp** [comment]
We should specify a strategy for non-settleable channels. Currently, the node will just shutdown without any indication of errors. Users should be aware of the fact that losing the channel state can lock the funds forever.


We could (sorted by difficulty)
1) shutdown the node and print informational message to the user
10) if settle fails, set a flag for such channel and do not attempt to settle it again
....
1000) try to bruteforce the balance hash (this can obviously work only for channels with an empty locksroot and low token decimals)

**Comment by hackaugusto** [comment]
> Users should be aware of the fact that losing the channel state can lock the funds forever.

No data is lost, the data is available through the WAL (although it's not straight forward to recover it).

> We should specify a strategy for non-settleable channels

We have a series of issues open to address the problem:

- Implement a queue of pending transactions #1950
- Check if the channel_identifier that requested it is old before making an on-chain transaction #1942
- Forbid on-chain actions which can race with a reorg #1965
- [META] Use more specific exceptions for the reasons a transactions may fail #1966

**Comment by pcppcp** [comment]
What I meant is a specific use case for a dumb user that does `rm -rf ~/.raiden`. Currently it renders the node unoperable - it'll try to settle with (0, 0, 0) after replaying `ChannelClosed` event.

`Invalid operations on the netting channel: settle with invalid balance proof data` from #1966 kind of a covers it

I am also thinking that comparing balance hash used to close/update channel with local node state could be used to decide whether to send actual `settle` transaction - it makes no sense to send tx that will 100 % fail.

**Comment by LefterisJP** [comment]
> What I meant is a specific use case for a dumb user that does rm -rf ~/.raiden

Understood. Yeah that is indeed an interesting use case to try and improve for (as we really can't do much apart from brute forcing as you said) but rather low priority. I already put it in [the docs](http://raiden-network.readthedocs.io/en/latest/overview_and_guide.html#requirements-for-safe-usage) that persistency of the DB is a requirement and people should not delete it.

>  that comparing balance hash used to close/update channel with local node state could be used to decide whether to send actual settle transaction -

Hmm yeah that would be a good idea

**Comment by pcppcp** [comment]
Fixed by #2006
